### PR TITLE
Add Google calendar event handler

### PR DESCRIPTION
**Pull request by chime3**
Closes #97

### CODE REVIEW

**Comment by chime3** [comment]
calendar-event-tests
calendar-tests

File: src/providers/google/calendar-event.ts [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+        };
+        this.emit('log', logEvent);
+        breakHit = SyncItemsBreak.TIMESTAMP;
+        break;
+      }
+
+      const insertedAt = new Date().toISOString();
+
+      const creator: Person = {
+        email: event.creator.email ?? 'info@example.com',
```
Comment (tahpot) [inline src/providers/google/calendar-event.ts:237]: We can't have `info@example.com` values

**Review by tahpot — CHANGES_REQUESTED** [review]
See comments

**Comment by tahpot** [comment]
Test isn't passing for me.

Screenshot 2024-09-30 at 6 36 01 PM

File: src/providers/google/calendar-event.ts [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+      let attendees: Person[] = []
+      if (event.attendees) {
+        attendees = event.attendees.filter(attendee => attendee.email) as Person[];
+      }
+
+      const attachments: CalendarAttachment[] = event.attachments as CalendarAttachment[];
+
+      results.push({
+        _id: this.buildItemId(eventId),
+        name: event.summary ?? 'No event title',
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/calendar-event.ts]: > `No title`

Removed

File: src/providers/google/calendar.ts [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            level: SyncProviderLogLevel.DEBUG,
+            message: `Invalid timezone for calendar ${calendarId}. Ignoring this calendar.`,
+        };
+        this.emit('log', logEvent);
+        continue;
+      }
+      
+      timeZone = CalendarHelpers.getUTCOffsetTimezone(timeZone);
+
+      const description = listItem.description ?? 'No description';
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/calendar.ts]: > Can these be blank, instead of `No xxx`?

Removed

File: src/providers/google/calendar.ts [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        const logEvent: SyncProviderLogEvent = {
+          level: SyncProviderLogLevel.DEBUG,
+          message: `Break ID hit (${breakId})`
+        };
+        this.emit('log', logEvent);
+        breakHit = SyncItemsBreak.ID;
+        break;
+      }
+
+      const summary = listItem.summary ?? 'No calendar title';
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/calendar.ts]: > `No title`

Removed

File: src/providers/google/calendar-event.ts [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        sourceApplication: this.getProviderApplicationUrl(),
+        sourceId: eventId,
+        calendarId: "primary",
+        start,
+        end,
+        creator,
+        organizer,
+        location: event.location ?? 'No location',
+        description: event.description ?? 'No description',
+        status: event.status ?? 'Unkown',
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/calendar-event.ts:230]: > What are the options for status?
In general, the options are "confirmed", "tentative", "cancelled", but optional

File: src/providers/google/calendar-event.ts [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        sourceData: event,
+        sourceApplication: this.getProviderApplicationUrl(),
+        sourceId: eventId,
+        calendarId: "primary",
+        start,
+        end,
+        creator,
+        organizer,
+        location: event.location ?? 'No location',
+        description: event.description ?? 'No description',
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/calendar-event.ts]: > Can we have `location`, `status` and `description` as empty values, instead of `No xxx`?

Fixed

File: src/providers/google/calendar-event.ts [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+      const attachments: CalendarAttachment[] = event.attachments as CalendarAttachment[];
+
+      results.push({
+        _id: this.buildItemId(eventId),
+        name: event.summary ?? 'No event title',
+        sourceAccountId: this.provider.getProviderId(),
+        sourceData: event,
+        sourceApplication: this.getProviderApplicationUrl(),
+        sourceId: eventId,
+        calendarId: "primary",
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/calendar-event.ts]: > This should be a reference to the `calendar._id` object associated with this event.

Fixed

File: src/providers/google/calendar-event.ts (hunk 1 above)
Comment (chime3) [reply src/providers/google/calendar-event.ts]: > We can't have `info@example.com` values

Removed

File: src/providers/google/interfaces.ts [hunk 8]
```diff
@@ -40,4 +40,27 @@ export enum YoutubeActivityType {
...
+  fileUrl?: string;   // URL of the file
+  title?: string;     // Title of the attachment
+  mimeType?: string;  // MIME type of the file
+  iconLink?: string;  // URL of the icon representing the file
+  fileId?: string;    // Unique identifier for the file
+}
+
+export interface GoogleCalendarHandlerConfig extends BaseHandlerConfig {
+  calendarLimit?: number;  // Max number of calendar per sync
+  eventsPerCalendarLimit?: number; // Max number of event to process in a calendar
```
Comment (tahpot) [inline src/providers/google/interfaces.ts:65]: Rename toto `calendarBatchSize` and `eventBatchSize` for consistency with other handlers.

File: src/serverconfig.example.json [hunk 9]
```diff
@@ -86,16 +87,20 @@
         "google": {
             "clientId": "",
             "clientSecret": "",
-            "batchSize": 50,
+            "batchSize": 30,
```
Comment (tahpot) [inline src/serverconfig.example.json:90]: Why was this changed to 30? I think it should be default of 50 for all handlers.

File: src/serverconfig.example.json [hunk 10]
```diff
@@ -86,16 +87,20 @@
...
             "sizeLimit": 10,
             "maxSyncLoops": 1,
-            "breakTimestamp": "2000-07-21T12:07:11.000Z",
             "handlers": {
                 "gmail": {
                     "batchSize": 200
                 },
                 "google-drive-document": {
-                    "batchSize": 50
+                    "batchSize": 2
```
Comment (tahpot) [inline src/serverconfig.example.json:98]: This should stay at 50

File: src/providers/google/calendar-event.ts [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+
+        if (!calendarId) {
+          this.emit("log", {
+            level: SyncProviderLogLevel.DEBUG,
+            message: `Invalid calendar ID. Ignoring this calendar.`,
+          });
+          continue;
+        }
+
+        const summary = calendar.summary ?? "No calendar title";
```
Comment (tahpot) [inline src/providers/google/calendar-event.ts:103]: Replace `No calendar title` with `No title`

File: src/providers/google/calendar-event.ts [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+    range: ItemsRange,
+    apiClient: calendar_v3.Calendar
+  ): Promise<SchemaEvent[]> {
+    const events: SchemaEvent[] = [];
+
+    let query: calendar_v3.Params$Resource$Events$List = {
+      calendarId: calendar.sourceId,
+      maxResults: this.config.eventsPerCalendarLimit,
+      singleEvents: true,
+      orderBy: "startTime"
```
Comment (tahpot) [inline src/providers/google/calendar-event.ts:155]: Google supports sorting by last updated, so we should use that so we automatically apply any changes to events.

File: src/providers/google/calendar-event.ts [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+    };
+
+    if (range.startId && !isNaN(Date.parse(range.startId))) {
```
Comment (tahpot) [inline src/providers/google/calendar-event.ts:158]: I would prefer if range.endId used pageToken from the Google API instead of a timestamp.

That is the exact placeholder of the next result set as specified by the Google API.

File: src/providers/google/calendar-event.ts [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+      query.timeMin = new Date(range.startId).toISOString();
+    }
+
+    if (range.endId && !isNaN(Date.parse(range.endId))) {
```
Comment (tahpot) [inline src/providers/google/calendar-event.ts:162]: `range.endId` should be the ID of the record that is at the end of the range and if it is hit, then the loop ends. No need to limit in the Google API query.

File: src/providers/google/calendar-event.ts [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+  }
+
+  private updateSyncPosition(
+    syncPosition: SyncHandlerPosition,
+    totalEvents: number,
+    calendarCount: number,
+  ) {
+    if (totalEvents === 0) {
+      syncPosition.status = SyncHandlerStatus.ENABLED;
+      syncPosition.syncMessage = "No new events found.";
```
Comment (tahpot) [inline src/providers/google/calendar-event.ts:384]: `Stopping. No results found.` for consistency with Gmail.

**Review by tahpot — CHANGES_REQUESTED** [review]

File: src/providers/google/calendar-event.ts [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+
+      events.push({
+        _id: this.buildItemId(eventId),
+        name: event.summary ?? "Unknown",
+        sourceAccountId: this.provider.getAccountId(),
+        sourceData: event,
+        sourceApplication: this.getProviderApplicationUrl(),
+        sourceId: eventId,
+        schema: CONFIG.verida.schemas.EVENT,
+        calendarId: calendar.sourceId ?? "primary",
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/calendar-event.ts:223]: Removed

File: src/providers/google/calendar-event.ts [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+  }
+
+  protected async buildCalendarList(): Promise<SchemaCalendar[]> {
+    const calendarClient = this.getCalendarClient();
+
+    let calendarList: SchemaCalendar[] = [];
+
+    let nextPageToken: string | undefined;
+    let query: calendar_v3.Params$Resource$Calendarlist$List = {
+      maxResults: MAX_BATCH_SIZE,  // Fetch in batches up to the max limit
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/calendar-event.ts:83]: Fixed

File: src/providers/google/interfaces.ts (hunk 8 above)
Comment (chime3) [reply src/providers/google/interfaces.ts:65]: Fixed

File: src/serverconfig.example.json (hunk 9 above)
Comment (chime3) [reply src/serverconfig.example.json:90]: It was for test, and reverted

File: src/providers/google/calendar-event.ts (hunk 11 above)
Comment (chime3) [reply src/providers/google/calendar-event.ts:103]: Fixed

File: src/providers/google/calendar-event.ts (hunk 12 above)
Comment (chime3) [reply src/providers/google/calendar-event.ts:155]: Done

File: src/providers/google/calendar-event.ts (hunk 13 above)
Comment (chime3) [reply src/providers/google/calendar-event.ts:158]: Done

File: src/providers/google/calendar-event.ts (hunk 14 above)
Comment (chime3) [reply src/providers/google/calendar-event.ts:162]: Fixed

File: src/providers/google/calendar-event.ts (hunk 15 above)
Comment (chime3) [reply src/providers/google/calendar-event.ts:384]: Fixed

**Comment by chime3** [comment]
### What's changed
- Replaced `timestamp` based item tracking with ID & `pageToken` tracking
- Modified common units for custom handler config
- Common unit tests cases are mainly checking basic item IDs 

### How to test
image

File: src/serverconfig.example.json (hunk 10 above)
Comment (tahpot) [reply src/serverconfig.example.json:101]: I've fixed this.

**Review by tahpot — CHANGES_REQUESTED** [review]
You can't make changes to `tests/common.tests.ts`. That is used by the other handler tests.

**Comment by chime3** [comment]
> You can't make changes to `tests/common.tests.ts`. That is used by the other handler tests.

Reverted and didn't use `runGenericTests` function of common unit tests, but used `buildTestObjects` and `checkItem` function.

Instead, added test cases in `calendar-event.tests.ts` file.
